crates/biome_js_analyze/src/lint/nursery/use_new_for_builtins.rs

```diff
@@ -0,0 +1,273 @@
+use crate::{services::semantic::Semantic, JsRuleAction};
+use biome_analyze::{
+    context::RuleContext, declare_rule, ActionCategory, Rule, RuleDiagnostic, RuleSource,
+};
+use biome_console::markup;
+use biome_diagnostics::Applicability;
+use biome_js_factory::make::{self, token_decorated_with_space};
+use biome_js_syntax::{
+    global_identifier, AnyJsExpression, JsCallExpression, JsNewExpression, JsSyntaxKind,
+};
+use biome_rowan::{chain_trivia_pieces, declare_node_union, AstNode, BatchMutationExt};
+
+declare_rule! {
+    /// Enforce the use of new for all builtins, except String, Number, Boolean, Symbol and BigInt.
+    ///
+    /// They work the same, but new should be preferred for consistency with other constructors.
+    /// Enforces the use of new for following builtins:
+    ///
+    /// - Object
+    /// - Array
+    /// - ArrayBuffer
+    /// - BigInt64Array
+    /// - BigUint64Array
+    /// - DataView
+    /// - Date
+    /// - Error
+    /// - Float32Array
+    /// - Float64Array
+    /// - Function
+    /// - Int8Array
+    /// - Int16Array
+    /// - Int32Array
+    /// - Map
+    /// - WeakMap
+    /// - Set
+    /// - WeakSet
+    /// - Promise
+    /// - RegExp
+    /// - Uint8Array
+    /// - Uint16Array
+    /// - Uint32Array
+    /// - Uint8ClampedArray
+    /// - SharedArrayBuffer
+    /// - Proxy
+    /// - WeakRef
+    /// - FinalizationRegistry
+    ///
+    /// Disallows the use of new for following builtins.
+    ///
+    /// - String
+    /// - Number
+    /// - Boolean
+    /// - Symbol
+    /// - BigInt
+    ///
+    /// > These should not use new as that would create object wrappers for the primitive values, which is not what you want. However, without new they can be useful for coercing a value to that type.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// const list = Array(10);
+    /// const now = Date();
+    /// const map = Map([
+    ///   ['foo', 'bar']
+    /// ]);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const list = new Array(10);
+    /// const now = new Date();
+    /// const map = new Map([
+    ///  ['foo', 'bar']
+    /// ]);
+    /// ```
+    ///
+    pub UseNewForBuiltins {
+        version: "next",
+        name: "useNewForBuiltins",
+        sources: &[RuleSource::EslintUnicorn("new-for-builtins")],
+        recommended: false,
+    }
+}
+
+const BUILTINS_REQUIRING_NEW: &[&str] = &[
```

Comment:
To increase the performance of the rule, I suggest you make the entries of this array **sorted**. Then, use the `binary_search` method.

Since entries need to be sorted, a test should be provided to possible regressions.

Here's an internal example: https://github.com/biomejs/biome/blob/main/crates%2Fbiome_js_analyze%2Fsrc%2Fglobals%2Fjavascript%2Fnode.rs#L164-L179